sys/unix/nleshared.cc

```diff
@@ -0,0 +1,707 @@
+
+#if defined(__linux__) && defined(__x86_64__)
+
+#include <elf.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <sys/mman.h>
+#include <dlfcn.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#include <string_view>
+#include <string>
+#include <cstdio>
+#include <stdexcept>
+#include <system_error>
+#include <vector>
+#include <cstddef>
+#include <cstring>
+#include <unordered_map>
+#include <mutex>
+
+struct NHLoader {
+  std::vector<std::byte> data;
+  Elf64_Ehdr* hdr;
+  size_t baseaddr;
+  size_t endaddr;
+  size_t size;
+  int fd;
+  void* ptr;
+  std::byte* sharedbase;
+  std::vector<std::pair<uint64_t, uint64_t>> relocations;
+  std::vector<std::pair<uint64_t, uint64_t>> writableRelocations;
+  size_t initfunc;
+  size_t finifunc;
+  size_t initarray;
+  size_t initarraysz;
+  size_t finiarray;
+  size_t finiarraysz;
+  NHLoader(std::string libPath, const std::unordered_map<std::string, void*>& overrides) {
+    FILE* f = fopen(libPath.c_str(), "rb");
+    if (!f) {
+      throw std::runtime_error("Failed to open '" + libPath + "' for reading");
+    }
+    fseek(f, 0, SEEK_END);
+    data.resize(ftell(f));
+    fseek(f, 0, SEEK_SET);
+    fread(data.data(), data.size(), 1, f);
+    fclose(f);
+
+    hdr = (Elf64_Ehdr*)data.data();
+    boundsCheck(hdr);
+
+    baseaddr = ~0;
+    endaddr = 0;
+    size = 0;
+
+    forPh([&](Elf64_Phdr* ph) {;
```

Comment:
I have a very vague notion of what this is doing: It's like a reimplementation of parts of Linux dynamic library loading. It interprets the data in the so in ELF format, creates manual copies ("forks"), and gives access to symbols a la `dlsym`.

I can't claim I understand much beyond that and perhaps a few comments might be good?

Reply:
I will try to outline the behavior in comments.